### DISCUSSION TITLE
Issue on n_parents using MOEAD

### DISCUSSION THREAD

**Discussion by MassimoValle** [anyoptimization/pymoo#511] [Q&A]
Hi, I am need of a help.

Running MOEA/D with MixedVariableSampling() and with MixedVariableMating() I get the following error:
`AttributeError: 'dict' object has no attribute 'n_parents'`

It refers to this line in the moead.py class [line 103]:
`P = self.selection.do(self.problem, pop, 1, self.mating.crossover.n_parents, neighbors=[self.neighbors[k]])`

The error occurs at the end of the first iteration, when the algorithm is preparing to perform the mating.

I can't figure out where I am wrong in the implementation and what parameter I am not setting or I am setting but incorrectly.
This is my implementation of the code:

            problem = MyProblem(fast=True)

            partitions = 3
            ref_dirs = get_reference_directions("das-dennis", 2, n_partitions=partitions)

            algorithm = MOEAD(
                ref_dirs= ref_dirs,
                sampling=MixedVariableSampling(),
                mating=MixedVariableMating(eliminate_duplicates=MixedVariableDuplicateElimination()),
                n_neighbors=12,
                prob_neighbor_mating= 0.7
            )

            res = minimize(
                problem,
                algorithm,
                ('n_gen', 30),
                seed=1,
                save_history=True,
                verbose=False
            )

I am pretty sure the problem is defined correctly since with NSGA2, NSGA3 and GA it works correctly.

Anyone can help me?

**Comment by blankjul** [comment]
`MOEAD` does not support the optimization of mixed variables. You will have to use `GA` or `NSGA2` as shown in the tutorial.
If you think `MOEAD` can be rewritten to make it work, please let me know or directly create a PR and I will have a look.

**Reply by MassimoValle** [reply]
Thank you for your reply.
May I ask from which version of pymoo the MOEA/D algorithm is no longer supported for mixed variable optimization?
This is because I have used this code in the past that used the optimization of mixed variables with MOEA/D and it worked correctly:

    from random import Random
    import numpy as np
    from pymoo.core.problem import ElementwiseProblem
    from pymoo.factory import get_sampling, get_crossover, get_mutation, get_reference_directions
    from pymoo.operators.mixed_variable_operator import MixedVariableSampling, MixedVariableMutation, MixedVariableCrossover
    from pymoo.factory import get_sampling, get_crossover, get_mutation
    from pymoo.optimize import minimize
    import sys
    from pymoo.core.problem import starmap_parallelized_eval
    from multiprocessing.pool import ThreadPool
    from pymoo.algorithms.moo.moead import MOEAD, ParallelMOEAD
    import random
    
    
    mask = ["int", "real", "int", "real"]
    
    sampling = MixedVariableSampling(mask, {
        "real": get_sampling("real_random"),
        "int": get_sampling("int_random"),
    })
    
    crossover = MixedVariableCrossover(mask, {
        "real": get_crossover("real_sbx"),
        "int": get_crossover("int_sbx")
    })
    
    mutation = MixedVariableMutation(mask, {
        "real": get_mutation("real_pm"),
        "int": get_mutation("int_pm")
    })
    
    
    class MyProblem(ElementwiseProblem):
    
        def __init__(self, fast = False, verbose = False):
            super().__init__(n_var= 4, n_obj = 2, n_constr = 0, xl = [1, 12.0, 2, 0.7], xu = [50, 30000.0, 11, 4.0])
            self.evaluationNumber = 0
            self.fast = fast
            self.verbose = verbose
            self.count = 0
    
        def _evaluate(self, x, out, *args, **kwargs):
            self.evaluationNumber = self.evaluationNumber + 1
    	    # other stuffs
    
    
    def main():
    
      problem = MyProblem(fast=fast, verbose=args.verbose)
      numb_obj = 2
      
      partitions = 3
      ref_dirs = get_reference_directions("das-dennis", numb_obj, n_partitions=partitions)
      algorithm = MOEAD(
          n_offsprings= None,
          sampling= sampling,
          crossover= crossover,
          mutation= mutation,
          ref_dirs= ref_dirs,
          n_neighbors=15,
          prob_neighbor_mating= 0.7
      )
      
      res = minimize(
                  problem,
                  algorithm,
                  seed= None,
                  save_history= True,
                  verbose= True
              )

**Reply by blankjul** [reply]
ohh interesting that it has even worked before (I did not know this because I have never tested it with MOEAD).
My best guess would be 0.6.0 or 0.5.0. If this is a minor fix I am happy to check a PR and merge it quickly to make it work again.
